src/Hooks.php

```diff
@@ -277,30 +280,35 @@ public function apply_filters($tag, $value)
     }
 
     $args = func_get_args();
-    array_shift($args);
+
+    array_shift($args); // drop tag.
 
     $this->__bump_action($tag, $args);
 
-    do {
-      foreach (current($this->filters[$tag]['_']) as $the_)
-      {
-        if (!is_null($the_['function']) && $the_['enabled'])
-        {
-          if (!is_null($the_['include_path']))
-          {
+    foreach ($this->filters[$tag]['_'] as $filters) {
+      foreach ($filters as $the_) {
+        if (isset($the_['function']) && $the_['enabled']) {
+          if (!is_null($the_['include_path'])) {
             include_once($the_['include_path']);
           }
 
-          $args[0] = $value;
-          $value = call_user_func_array($the_['function'], $args);
+          $args = call_user_func_array($the_['function'], $args);
+
+          if(!is_array($args)) { // must be an array for next call_user_func_array(___,$args)
+            $args = [$args];
+          }
         }
       }
     }
-    while (next($this->filters[$tag]['_']) !== false);
+
+    if(sizeof($args) == 1) { // assume we should really be passing back a single value (e.g. $blah = apply_filter('bar', $blah))
+      return $args[0];
```

Comment:
if you return before `array_pop($this->filters_stack);` it will probably fail somewhere

But as my other comments, this condition has to be removed.